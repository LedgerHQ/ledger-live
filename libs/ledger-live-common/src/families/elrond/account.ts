import invariant from "invariant";
import type { Operation } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
import type { Unit } from "@ledgerhq/types-cryptoassets";
import type { ElrondAccount } from "./types";

function formatAccountSpecifics(account: ElrondAccount): string {
  const { elrondResources } = account;
  invariant(elrondResources, "elrond account expected");
  const unit = getAccountUnit(account);
  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };
  let str = " ";

  if (account.spendableBalance) {
    str +=
      formatCurrencyUnit(unit, account.spendableBalance, formatConfig) +
      " spendable. ";
  } else {
    str += " 0 spendable.";
  }

  if (elrondResources && elrondResources.nonce) {
    str += "\n nonce : " + elrondResources.nonce;
  }

  return str;
}

function formatOperationSpecifics(
  _op: Operation,
  _unit: Unit | null | undefined
): string {
  return "";
}

export default {
  formatAccountSpecifics,
  formatOperationSpecifics,
};
