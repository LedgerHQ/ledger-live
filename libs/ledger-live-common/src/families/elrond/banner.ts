import type { ElrondAccount, ElrondProvider } from "./types";
import { ElrondDelegation, ElrondPreloadData } from "./types";
import { ELROND_LEDGER_VALIDATOR_ADDRESS } from "./constants";
import { hasMinimumDelegableBalance } from "@ledgerhq/coin-elrond/helpers/hasMinimumDelegableBalance";

interface AccountBannerHiddenState {
  bannerType: "hidden";
}

interface AccountBannerDelegateState {
  bannerType: "delegate";
}

interface AccountBannerRedelegateState {
  bannerType: "redelegate";
  worstDelegation: ElrondDelegation;
}

export type AccountBannerState =
  | AccountBannerDelegateState
  | AccountBannerRedelegateState
  | AccountBannerHiddenState;

function getWorstValidator(
  delegations: ElrondDelegation[],
  validators: ElrondProvider[],
  ledgerValidator: ElrondProvider,
) {
  return delegations.reduce((worstValidator, delegation) => {
    const validator = validators.find(validator => validator.contract === delegation.contract);
    if (validator && validator.aprValue < worstValidator.aprValue) {
      return validator;
    }
    return worstValidator;
  }, ledgerValidator);
}

export function getAccountBannerState(
  account: ElrondAccount,
  elrondPreloadData: ElrondPreloadData,
): AccountBannerState {
  const { validators } = elrondPreloadData;
  const elrondResources = account.elrondResources
    ? account.elrondResources
    : { delegations: [] as ElrondDelegation[] };

  const hasDelegations = elrondResources.delegations.length > 0;
  const ledgerValidator = validators.find(
    validator => validator.contract === ELROND_LEDGER_VALIDATOR_ADDRESS,
  );

  // // if Ledger doesn't provide validator, we don't display banner
  if (!ledgerValidator) {
    return {
      bannerType: "hidden",
    };
  }

  if (hasDelegations) {
    const worstValidator: ElrondProvider = getWorstValidator(
      elrondResources.delegations,
      validators,
      ledgerValidator,
    );

    if (worstValidator.contract !== ledgerValidator.contract) {
      const worstDelegation = elrondResources.delegations.find(
        delegation => delegation.contract === worstValidator.contract,
      );
      if (!worstDelegation) return { bannerType: "hidden" };

      return {
        bannerType: "redelegate",
        worstDelegation,
      };
    }
  }

  if (hasMinimumDelegableBalance(account)) {
    return {
      bannerType: "delegate",
    };
  }

  return {
    bannerType: "hidden",
  };
}
