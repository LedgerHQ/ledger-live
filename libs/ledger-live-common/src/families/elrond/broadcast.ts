import type { AccountBridge } from "@ledgerhq/types-live";
import { patchOperationWithHash } from "../../operation";
import { broadcastTransaction } from "./api";
import { Transaction } from "./types";

/**
 * Broadcast the signed transaction
 * @param {signature: string, operation: string} signedOperation
 */
export const broadcast: AccountBridge<Transaction>["broadcast"] = async ({ signedOperation }) => {
  const hash = await broadcastTransaction(signedOperation);

  return patchOperationWithHash(signedOperation.operation, hash);
};

export default broadcast;
