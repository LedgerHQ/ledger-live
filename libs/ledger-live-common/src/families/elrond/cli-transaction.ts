import type { AccountLike } from "@ledgerhq/types-live";
import invariant from "invariant";
import flatMap from "lodash/flatMap";
import type { Transaction } from "../../generated/types";
import type { ElrondAccount } from "./types";
const options = [
  {
    name: "mode",
    type: String,
    desc: "mode of transaction: send",
  },
];

function inferTransactions(
  transactions: Array<{
    account: AccountLike;
    transaction: Transaction;
  }>,
  _opts: Record<string, any>
): Transaction[] {
  return flatMap(transactions, ({ transaction, account }) => {
    invariant(transaction.family === "elrond", "elrond family");

    if (account.type === "Account") {
      invariant(
        (account as ElrondAccount).elrondResources,
        "unactivated account"
      );
    }

    transaction.family = "elrond";

    return transaction;
  });
}

export default {
  options,
  inferTransactions,
};
