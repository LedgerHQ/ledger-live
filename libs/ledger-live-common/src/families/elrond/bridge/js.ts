import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import { getPreloadStrategy, preload, hydrate } from "../preload";
import { sync, scanAccounts } from "../js-synchronisation";
import { createTransaction, prepareTransaction } from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { assignFromAccountRaw, assignToAccountRaw } from "../serialization";

const receive = makeAccountBridgeReceive();
const currencyBridge: CurrencyBridge = {
  getPreloadStrategy,
  preload,
  hydrate,
  scanAccounts,
};
const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction: defaultUpdateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
  assignFromAccountRaw,
  assignToAccountRaw,
};
export default {
  currencyBridge,
  accountBridge,
};
