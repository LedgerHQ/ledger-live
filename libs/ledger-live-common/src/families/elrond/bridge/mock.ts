// TODO: update path by moving mockHelpers to coin-framework

import { BigNumber } from "bignumber.js";
import { NotEnoughBalance, RecipientRequired, InvalidAddress, FeeTooHigh } from "@ledgerhq/errors";
import type { ElrondAccount, Transaction } from "@ledgerhq/coin-elrond/types";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import {
  makeAccountBridgeReceive,
  scanAccounts,
  signOperation,
  broadcast,
  sync,
  isInvalidRecipient,
} from "../../../bridge/mockHelpers";
import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import { getMainAccount } from "@ledgerhq/coin-framework/account/index";
import { assignToAccountRaw, assignFromAccountRaw } from "@ledgerhq/coin-elrond/serialization";

const receive = makeAccountBridgeReceive();

const defaultGetFees = (a, t) => t.fees || new BigNumber(0);

const createTransaction = (): Transaction => ({
  family: "elrond",
  mode: "send",
  amount: new BigNumber(0),
  recipient: "",
  fees: null,
  useAllAmount: false,
  gasLimit: 0,
});

const estimateMaxSpendable = ({ account, parentAccount, transaction }) => {
  const mainAccount = getMainAccount(account, parentAccount);
  const estimatedFees = transaction
    ? defaultGetFees(mainAccount, transaction)
    : new BigNumber(5000);
  return Promise.resolve(BigNumber.max(0, account.balance.minus(estimatedFees)));
};

const getTransactionStatus = (account: ElrondAccount, transaction: Transaction) => {
  const errors: Record<string, Error> = {};
  const warnings: Record<string, Error> = {};
  const useAllAmount = !!transaction.useAllAmount;
  const estimatedFees = defaultGetFees(account, transaction);
  const totalSpent = useAllAmount
    ? account.balance
    : new BigNumber(transaction.amount).plus(estimatedFees);
  const amount = useAllAmount
    ? account.balance.minus(estimatedFees)
    : new BigNumber(transaction.amount);

  if (amount.gt(0) && estimatedFees.times(10).gt(amount)) {
    warnings.feeTooHigh = new FeeTooHigh();
  }

  // Fill up transaction errors
  if (totalSpent.gt(account.balance)) {
    errors.amount = new NotEnoughBalance();
  }

  // Fill up recipient errors
  if (!transaction.recipient) {
    errors.recipient = new RecipientRequired("");
  } else if (isInvalidRecipient(transaction.recipient)) {
    errors.recipient = new InvalidAddress("", {
      currencyName: account.currency.name,
    });
  }

  return Promise.resolve({
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  });
};

const prepareTransaction = async (a, t) => {
  if (!t.fees) {
    return { ...t, fees: new BigNumber(500) };
  }

  return t;
};

const accountBridge: AccountBridge<Transaction, any> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction: defaultUpdateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  assignToAccountRaw,
  assignFromAccountRaw,
  signOperation,
  broadcast,
};
const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: () => {
    return Promise.resolve({});
  },
  hydrate: () => {},
};
export default {
  currencyBridge,
  accountBridge,
};
