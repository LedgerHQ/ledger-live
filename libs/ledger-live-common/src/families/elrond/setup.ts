// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-elrond/bridge/js";
import makeCliTools from "@ledgerhq/coin-elrond/cli-transaction";
import elrondResolver from "@ledgerhq/coin-elrond/hw-getAddress";
import { ElrondAccount, Transaction, TransactionStatus } from "@ledgerhq/coin-elrond/types";
import Elrond from "@ledgerhq/hw-app-elrond";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Elrond> = (transport: Transport) => {
  return new Elrond(transport);
};

const bridge: Bridge<Transaction, ElrondAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, elrondResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
