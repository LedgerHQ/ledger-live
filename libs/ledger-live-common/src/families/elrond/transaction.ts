import type { Transaction, TransactionRaw } from "./types";
import { BigNumber } from "bignumber.js";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon as fromTransactionStatusRaw,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon as toTransactionStatusRaw,
} from "@ledgerhq/coin-framework/transaction/common";
import type { Account } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
export const formatTransaction = (
  { mode, amount, recipient, useAllAmount, subAccountId }: Transaction,
  mainAccount: Account
): string => {
  const account =
    (subAccountId &&
      (mainAccount.subAccounts || []).find((a) => a.id === subAccountId)) ||
    mainAccount;
  return `
${mode.toUpperCase()} ${
    useAllAmount
      ? "MAX"
      : amount.isZero()
      ? ""
      : " " +
        formatCurrencyUnit(getAccountUnit(account), amount, {
          showCode: true,
          disableRounding: true,
        })
  }${recipient ? `\nTO ${recipient}` : ""}`;
};
export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);
  return {
    ...common,
    family: tr.family,
    mode: tr.mode,
    fees: tr.fees ? new BigNumber(tr.fees) : null,
    data: tr.data,
    gasLimit: tr.gasLimit,
  };
};

export const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);
  return {
    ...common,
    family: t.family,
    mode: t.mode,
    fees: t.fees?.toString() || null,
    data: t.data,
    gasLimit: t.gasLimit,
  };
};

export default {
  formatTransaction,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
  formatTransactionStatus,
};
