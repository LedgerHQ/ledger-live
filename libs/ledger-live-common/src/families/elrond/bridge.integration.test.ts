import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";

import { fromTransactionRaw } from "../elrond/transaction";
import { BigNumber } from "bignumber.js";
import {
  InvalidAddressBecauseDestinationIsAlsoSource,
  NotEnoughBalance,
  InvalidAddress,
} from "@ledgerhq/errors";
import type { DatasetTest, CurrenciesData } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
const TEST_ADDRESS =
  "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2";

const elrond: CurrenciesData<Transaction> = {
  scanAccounts: [
    {
      name: "elrond seed 1",
      apdus: `
    => ed030000080000000080000000
    <= 3e6572643176676670336737617a716a783477736d7474373036376d306c3632763370736d717a7232346a36787679776a32746c7a3067657376797a7371329000
    => ed030000080000000080000000
    <= 3e6572643176676670336737617a716a783477736d7474373036376d306c3632763370736d717a7232346a36787679776a32746c7a3067657376797a7371329000
    => ed030000080000000080000001
    <= 3e65726431706d33676a65326c6d643576796c6463767579366a7078676465347261706a70756a76756a6b65653661376a77327a6d6834747172653739367a9000
    `,
    },
  ],
  accounts: [
    {
      raw: {
        id: `js:2:elrond:${TEST_ADDRESS}:`,
        seedIdentifier: `${TEST_ADDRESS}`,
        name: "Elrond 1",
        derivationMode: "",
        index: 0,
        freshAddress: `${TEST_ADDRESS}`,
        freshAddressPath: "44'/508'/0'/0'/0'",
        freshAddresses: [],
        blockHeight: 0,
        operations: [],
        pendingOperations: [],
        currencyId: "elrond",
        unitMagnitude: 18,
        lastSyncDate: "",
        balance: "299569965",
      },
      transactions: [
        {
          name: "recipient and sender must not be the same",
          transaction: fromTransactionRaw({
            family: "elrond",
            recipient: `${TEST_ADDRESS}`,
            amount: "100000000",
            mode: "send",
            fees: null,
          }),
          expectedStatus: {
            amount: new BigNumber("100000000"),
            errors: {
              recipient: new InvalidAddressBecauseDestinationIsAlsoSource(),
            },
            warnings: {},
          },
        },
        {
          name: "Not a valid address",
          transaction: fromTransactionRaw({
            family: "elrond",
            recipient: "elrondinv",
            amount: "100000000",
            mode: "send",
            fees: null,
          }),
          expectedStatus: {
            errors: {
              recipient: new InvalidAddress(),
            },
            warnings: {},
          },
        },
        {
          name: "Not enough balance",
          transaction: fromTransactionRaw({
            family: "elrond",
            recipient:
              "erd1frj909pfums4m8aza596595l9pl56crwdj077vs2aqcw6ynl28wsfkw9rd",
            amount: "1000000000000000000000000",
            mode: "send",
            fees: null,
          }),
          expectedStatus: {
            errors: {
              amount: new NotEnoughBalance(),
            },
            warnings: {},
          },
        },
      ],
    },
  ],
};

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    elrond,
  },
};

testBridge(dataset);
