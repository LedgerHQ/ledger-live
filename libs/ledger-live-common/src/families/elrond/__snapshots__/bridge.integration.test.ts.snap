// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elrond currency bridge scanAccounts elrond seed 1 1`] = `
Array [
  Object {
    "balance": "180162251239346240",
    "currencyId": "elrond",
    "derivationMode": "",
    "elrondResources": Object {
      "nonce": 3,
    },
    "freshAddress": "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
    "freshAddressPath": "44'/508'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
        "derivationPath": "44'/508'/0'/0/0",
      },
    ],
    "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
    "index": 0,
    "name": "Elrond 1",
    "nfts": undefined,
    "operationsCount": 6,
    "pendingOperations": Array [],
    "seedIdentifier": "",
    "spendableBalance": "180162251239346240",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": true,
  },
  Object {
    "balance": "0",
    "currencyId": "elrond",
    "derivationMode": "",
    "elrondResources": Object {
      "nonce": 0,
    },
    "freshAddress": "erd1pm3gje2lmd5vyldcvuy6jpxgde4rapjpujvujkee6a7jw2zmh4tqre796z",
    "freshAddressPath": "44'/508'/1'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "erd1pm3gje2lmd5vyldcvuy6jpxgde4rapjpujvujkee6a7jw2zmh4tqre796z",
        "derivationPath": "44'/508'/1'/0/0",
      },
    ],
    "id": "js:2:elrond:erd1pm3gje2lmd5vyldcvuy6jpxgde4rapjpujvujkee6a7jw2zmh4tqre796z:",
    "index": 1,
    "name": "Elrond 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": false,
  },
]
`;

exports[`elrond currency bridge scanAccounts elrond seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
      "blockHash": "6180412daff9134648325c6dcb035657c444e4d930f9e864bfa29f4e5c6e262d",
      "blockHeight": 5932298,
      "contract": undefined,
      "extra": Object {},
      "fee": "50000000000000",
      "hasFailed": false,
      "hash": "016af2b27940612e2b234df13a47b84e22d90694ddd65bc937526dd31b2c397c",
      "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:-016af2b27940612e2b234df13a47b84e22d90694ddd65bc937526dd31b2c397c-IN",
      "operator": undefined,
      "recipients": Array [
        "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
      ],
      "senders": Array [
        "erd1khl4mll00m6tgvcpmhf9tu05s9cfszruw77fk4dfmzyn7r2h6kgqdtsj7u",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "1000000000000000",
    },
    Object {
      "accountId": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
      "blockHash": "c106aceec57695d2ce2031c6546ad1c6816ead5503cd15444b76adca8de52a2e",
      "blockHeight": 5932325,
      "contract": undefined,
      "extra": Object {},
      "fee": "50000000000000",
      "hasFailed": false,
      "hash": "5ca1c624f3e51a81b9b76a08b45d6f2d8a0d4547983b924412afc72e21ba771a",
      "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:-5ca1c624f3e51a81b9b76a08b45d6f2d8a0d4547983b924412afc72e21ba771a-OUT",
      "operator": undefined,
      "recipients": Array [
        "erd1khl4mll00m6tgvcpmhf9tu05s9cfszruw77fk4dfmzyn7r2h6kgqdtsj7u",
      ],
      "senders": Array [
        "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 2,
      "type": "OUT",
      "value": "150000000000000",
    },
    Object {
      "accountId": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
      "blockHash": "279b51d3cf81495ce82cb72d9c9d5a854409e23f25c2553efec6084c6ccb13c8",
      "blockHeight": 5010572,
      "contract": undefined,
      "extra": Object {},
      "fee": "50000000000000",
      "hasFailed": false,
      "hash": "976d34a592aa5421a19c05fb68463ee3bf35801a8ada608f023843f0c9b0e07b",
      "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:-976d34a592aa5421a19c05fb68463ee3bf35801a8ada608f023843f0c9b0e07b-OUT",
      "operator": undefined,
      "recipients": Array [
        "erd1frj909pfums4m8aza596595l9pl56crwdj077vs2aqcw6ynl28wsfkw9rd",
      ],
      "senders": Array [
        "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 0,
      "type": "OUT",
      "value": "50000000000000",
    },
    Object {
      "accountId": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
      "blockHash": "da2121fa3f54786a2ac4739a31d7999538a87ffb0d3ed9ceedc2771350181eb5",
      "blockHeight": 5010620,
      "contract": undefined,
      "extra": Object {},
      "fee": "50000000000000",
      "hasFailed": false,
      "hash": "c0615e1eb10ed98fb0e0bcbb28588b3937ff1e5d80cb8ad0112a3380f36dbf6a",
      "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:-c0615e1eb10ed98fb0e0bcbb28588b3937ff1e5d80cb8ad0112a3380f36dbf6a-OUT",
      "operator": undefined,
      "recipients": Array [
        "erd1frj909pfums4m8aza596595l9pl56crwdj077vs2aqcw6ynl28wsfkw9rd",
      ],
      "senders": Array [
        "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 1,
      "type": "OUT",
      "value": "9950000000000000",
    },
    Object {
      "accountId": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
      "blockHash": "1b97c0f5da6c83e4dc6ae280cac7fe4b002c0274eb050793b2241cb823ef4551",
      "blockHeight": 5010539,
      "contract": undefined,
      "extra": Object {},
      "fee": "56000000000000",
      "hasFailed": false,
      "hash": "dbcc80e99b9e45fd4973e5e8a6b283d1df3ace0cd424b7f4dfd59eaac31bd54d",
      "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:-dbcc80e99b9e45fd4973e5e8a6b283d1df3ace0cd424b7f4dfd59eaac31bd54d-IN",
      "operator": undefined,
      "recipients": Array [
        "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
      ],
      "senders": Array [
        "erd1frj909pfums4m8aza596595l9pl56crwdj077vs2aqcw6ynl28wsfkw9rd",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "10000000000000000",
    },
    Object {
      "accountId": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:",
      "blockHash": "8f76929839d1475ee44570085198e2555df1fa4848ccd9e8ed25e68c810de135",
      "blockHeight": 7620524,
      "contract": undefined,
      "extra": Object {},
      "fee": "50000000000000",
      "hasFailed": false,
      "hash": "f12fdcf0d5ae80ed7f8c447faca6417dffa5ad1dc359681b9c19f706acd3812f",
      "id": "js:2:elrond:erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2:-f12fdcf0d5ae80ed7f8c447faca6417dffa5ad1dc359681b9c19f706acd3812f-IN",
      "operator": undefined,
      "recipients": Array [
        "erd1vgfp3g7azqjx4wsmtt7067m0l62v3psmqzr24j6xvywj2tlz0gesvyzsq2",
      ],
      "senders": Array [
        "erd1wh9c0sjr2xn8hzf02lwwcr4jk2s84tat9ud2kaq6zr7xzpvl9l5q8awmex",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "179312251239346240",
    },
  ],
  Array [],
]
`;
